keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+
+A `defaultValidations` field will allow policy definition authors to define
+expressions which are evaluated when the policy definition matches a resource
+but no policy configurations match that resource.
+
+Behavior:
+
+- `defaultValidations` do not have access to params.
+- `spec.paramSource` is optional only when `defaultValidations` are present
+- (`spec.paramSource` and `spec.validations` must both be either present or absent)
+- When a policy has `defaultValidations`, it must also declare the enforcement
+  level of those validations
+
+A "singleton" (aka standalone) policy can be defined as:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+...
+spec:
+  match: ...
+  defaultValidations:
+  - expression: "object.spec.replicas < 100"
+    enforcement: Deny
+    
+```
+
+"Principle of least privilege" policies are also possible:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+...
+spec:
+  paramSource:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  match:
+    ...
+  validations:
+    expression: "object.spec.replicas <= params.maxReplicas"
+    name: max-replicas
+  defaultValidations:
+  - expression: "object.spec.replicas <= 5"
+    enforcement: [Deny]
+```
+
+Reporting/debugging/analysis implications:
+
+- Violations for non-singleton policies will always be reported using a {policy
+  definition, policy binding} identifier pair. To be consistent with this,
+  singleton policies can be use a {policy definition, policy definition}
+  identifier pair. This is a bit verbose but keeps reporting consistent and
+  makes tracing a violation back to resources that produced it straight-forward.
+
+### Phase 2
+
+All these capabilities are required before Beta, but will not be implemented
+in the first alpha release of this enhancement due to the size and complexity
+of this enhancement.
+
+#### Namespace scoped policy binding
+
+For phase 1, policy bindings were only allowed to be cluster scoped. We can
+support namespace scoped policy bindings as follows:
+
+- Add a `NamespacePolicyBinding` resource.
+- If the parameter resource is namespace scoped, it implicitly matches
+  resources only in the namespace it is in, but may further constrain what
+  resources it matches with additional match criteria.
+
+Benefits: Allows policy of a namespace to be controlled from within the
+namespace. As an example, ResourceQuota works this way.
+
+Details to consider:
+
+- Should a policy support both cluster scoped and namespace scoped binding? If
+  so how? It would need two different parameter CRDs (since a CRD must either be
+  cluster scoped or namespace scoped, not both).
+
+#### CEL Expression Composition
+
+##### Variables
+
+<<[UNRESOLVED jpbetz, ?? ]>>
+cel-policy-template "terms" are very convenient. But we don't have consensus on
+if they are sufficiently essential. Need to either agree on something or drop
+this.
+<<[/UNRESOLVED]>>
```

Comment:
Benefits of variables:

* Simpliify computation by evaluating complex expressions once and storing the result in a variable for reuse
* Avoid the need to duplicate brittle and repetitive logic, such as field existence checks
* May obviate the need for the "scope" section below
* Avoids copy/pasting subexpressions in CEL expressions (I don't believe CEL lets you define variables?)

There might be more, but the above seem pretty persuasive to me.

Reply:
I think the readability arguments are sufficient to warrant the introduction of terms. There is some added complexity carried with doing so in that these features aren't part of the standard CEL-Go library, but can be orchestrated outside it.

Reply:
I think that `scopes` should be omitted in favor of bind variables as, they're nice, but basically a way of removing the `all`, `exists` macro from the expression. In other words the scope replicates existing CEL behavior, but the term does not.

Reply:
I think I'm sold on this. I've dropped scopes.

Any strong preference for what we call them? `variables`, `terms`, something else?